pkg/vm/vm.go

```diff
@@ -22,37 +27,342 @@ type VirtualMachineConfig struct {
+	}
+
+	nextPc, err := vm.updatePc(instruction, dstCell, op1Cell, res)
+	if err != nil {
+		return err
+	}
+
+	nextAp, err := vm.updateAp(instruction, res)
+	if err != nil {
+		return err
+	}
+
+	nextFp, err := vm.updateFp(instruction, dstCell)
+	if err != nil {
+		return err
+	}
+
+	vm.Context.Pc = nextPc
+	vm.Context.Ap = nextAp
+	vm.Context.Fp = nextFp
+
 	return nil
 }
 
-func (vm *VirtualMachine) RunHint(instruction *Instruction) error {
+func (vm *VirtualMachine) RunHint() error {
 	return nil
 }
+
+func (vm *VirtualMachine) getCellDst(instruction *Instruction) (*mem.Cell, error) {
+	var dstRegister uint64
+	if instruction.DstRegister == Ap {
+		dstRegister = vm.Context.Ap
+	} else {
+		dstRegister = vm.Context.Fp
+	}
+
+	// todo(rodro): fix this math
+	return vm.MemoryManager.Memory.Peek(executionSegment, dstRegister+uint64(instruction.OffDest))
+}
+
+func (vm *VirtualMachine) getCellOp0(instruction *Instruction) (*mem.Cell, error) {
+	var op0Register uint64
+	if instruction.Op0Register == Ap {
+		op0Register = vm.Context.Ap
+	} else {
+		op0Register = vm.Context.Fp
+	}
+	// todo(rodro): fix this math
+	offset := op0Register + uint64(instruction.OffOp0)
+	return vm.MemoryManager.Memory.Peek(executionSegment, offset)
+}
+
+func (vm *VirtualMachine) getCellOp1(instruction *Instruction, op0Cell *mem.Cell) (*mem.Cell, error) {
+	var op1Address *mem.MemoryAddress
+	switch instruction.Op1Source {
+	case Op0:
+		// in this case Op0 is being used as an address, and must be of unwrapped as is
+		op0Address, err := op0Cell.Read().ToMemoryAddress()
+		if err != nil {
+			return nil, fmt.Errorf("expected op0 to be an address: %w", err)
+		}
+		op1Address = mem.CreateMemoryAddress(op0Address.SegmentIndex, op0Address.Offset)
+	case Imm:
+		op1Address = mem.CreateMemoryAddress(programSegment, vm.Context.Pc)
+	case FpPlusOffOp1:
+		op1Address = mem.CreateMemoryAddress(executionSegment, vm.Context.Fp)
+	case ApPlusOffOp1:
+		op1Address = mem.CreateMemoryAddress(executionSegment, vm.Context.Ap)
+	}
+	// todo(rodro): fix this math
+	op1Address.Offset += uint64(instruction.OffOp1)
+
+	return vm.MemoryManager.Memory.PeekFromAddress(op1Address)
+}
+
+// when there is an assertion with a substraction or division like : x = y - z
+// the compiler treats it as y = x + z. This means that the VM knows the
+// dstCell value and either op0Cell xor op1Cell. This function infers the
+// unknow operand as well as the `res` auxiliar value
+func (vm *VirtualMachine) inferOperand(
+	instruction *Instruction, dstCell *mem.Cell, op0Cell *mem.Cell, op1Cell *mem.Cell,
+) (*mem.MemoryValue, error) {
+	if instruction.Opcode != AssertEq ||
+		(instruction.Res != AddOperands && instruction.Res != MulOperands) ||
+		(op0Cell.Accessed && op1Cell.Accessed) {
+		return nil, nil
+	}
+	if !dstCell.Accessed {
+		return nil, fmt.Errorf("impossible to infer unknown operand, dst cell is unknown too")
+	}
+
+	var knownOpCell *mem.Cell
+	var unknownOpCell *mem.Cell
+	if op0Cell.Accessed {
```

Comment:
What if dst is accessed op0 and op1 not?

Reply:
Yes, I was thinking on that. It should error, because calculations wouldn't be able to get completed. I think it's part of a larger problem on the logic side. When you read from an unaccessed Cell it should return an error instead of silently assigning a default zero value as per the other VMs implementation. What do you think?